### ISSUE TITLE
Wrong counter value when accessed from inside floating figure

### ISSUE THREAD

**Issue by EpicEricEE** [typst/typst#4966]
### Description

_This is a regression introduced with #4212_.

The issue is only concerning floating figures, i.e. where `placement` is not `none`.

When accessing the current figure counter value from inside a figure's body or caption, the counter seems to not yet have been updated. For example:

<table>
<tr>
<th>Code</th>
<th>Output on 0.11.1</th>
<th>Output on main</th>
</tr>

<tr>
<td>

```typ
#figure(
  context square(
    counter(figure.where(kind: image)).display()
  ),
  placement: bottom,
  caption: [A Square]
)
```

</td>

<td>

![image](https://github.com/user-attachments/assets/385bf1aa-cda7-4874-a8b0-114201d9ae22)

</td>

<td>

![image](https://github.com/user-attachments/assets/507daefa-44b0-40c4-87c2-8079b16e8f70)

</td>
</tr>

</table>

Possibly related to #4322.

### Operating system

Windows

### Typst version

- [X] I am using the latest version of Typst

**Comment by quachpas** [comment]
@EpicEricEE Are you able to reproduce this issue on 0.12-rc1 (d5b1f4a6f)? I am unable to reproduce this on either typst.app or locally.

**Comment by EpicEricEE** [comment]
I can't either after #5017, though an issue remains when there are multiple floating figures, as they now all show the same number, which was not the case in 0.11.1:

<table>

<tr>
<th>Code</th>
<th>0.11.1</th>
<th>0.12.0-rc1</th>
</tr>

<tr>
<td>

```typ
#let fig = figure(
  context square({
    counter(figure.where(kind: image)).display()
  }),
  placement: bottom,
  caption: [A Square]
)

#fig
#fig
```
</td>
<td>

![image](https://github.com/user-attachments/assets/53ec3800-fe17-418e-8e62-9552ecafe9fd)

</td>
<td>

![image](https://github.com/user-attachments/assets/aaf727ab-24db-45fd-ad59-8a471f0ddca9)

</td>
</tr>

</table>

**Comment by laurmaedje** [comment]
You can reproduce it by using `placement: top` in the original example. The reason being that with top placement, the figure ends up logically before the page contents and with bottom placement, logically after.

**Comment by laurmaedje** [comment]
This and https://github.com/typst/typst/issues/4359 are kind of at odds with each other. To fix #4966 requires to treat the figure content as if it is logically at the in-flow position where the figure is anchored. To fix #4359 requires to change the location of the figure itself to be at the top or bottom of the page, which breaks counters.
